lib/Dialect/FIRRTL/Ops.cpp

```diff
@@ -74,6 +75,118 @@ static LogicalResult verifyCircuitOp(CircuitOp &circuit) {
+              << "previous module declared here";
+          return failure();
+        }
+
+        // Find an optional extmodule with a defname collision. Update
+        // the defnameMap if this is the first extmodule with that
+        // defname or if the current extmodule takes no parameters and
+        // the collision does. The latter condition improves later
+        // extmodule verification as checking against a parameterless
+        // module is stricter.
+        auto collidingExtModuleOpt =
+            Optional<FExtModuleOp>(llvm::NoneType::None);
+        {
+          auto value = defnameMap.find(defname.getValue());
+          if (value != defnameMap.end()) {
+            collidingExtModuleOpt = Optional<FExtModuleOp>(value->getValue());
+            if (value->getValue().parameters() && !extModule.parameters())
+              defnameMap.insert_or_assign(defname.getValue(), extModule);
+          } else {
+            defnameMap.insert_or_assign(defname.getValue(), extModule);
+          }
+        }
+
+        // Go to next extmodule if no extmodule with the same defname
+        // was found.
+        if (!collidingExtModuleOpt)
+          continue;
+
+        auto collidingExtModule = collidingExtModuleOpt.getValue();
+
+        // Check that the number of ports is exactly the same.
+        auto ports = new SmallVector<std::pair<StringAttr, FIRRTLType>, 8>;
+        auto collidingPorts =
+            new SmallVector<std::pair<StringAttr, FIRRTLType>, 8>;
+        extModule.getPortInfo(*ports);
+        collidingExtModule.getPortInfo(*collidingPorts);
+        if (ports->size() != collidingPorts->size()) {
+          auto diag = op.emitOpError()
+                      << "with 'defname' attribute " << defname << " has "
+                      << ports->size()
+                      << " ports which is different from a previously defined "
+                         "extmodule with the same 'defname' which has "
+                      << collidingPorts->size() << " ports";
+          diag.attachNote(collidingExtModule.getLoc())
+              << "previous extmodule definition occurred here";
+          return failure();
+        }
+
+        // Check that ports match for name and type. Since parameters
+        // *might* affect widths, ignore widths if either module has
+        // parameters. Note that this allows for misdetections, but
+        // has zero false positives.
+        for (auto p : llvm::zip(*ports, *collidingPorts)) {
+          auto a_name = std::get<0>(std::get<0>(p));
+          auto b_name = std::get<0>(std::get<1>(p));
+          FIRRTLType a_type, b_type;
+          if (extModule.parameters() || collidingExtModule.parameters()) {
+            a_type = std::get<1>(std::get<0>(p)).getWidthlessType();
+            b_type = std::get<1>(std::get<1>(p)).getWidthlessType();
+          } else {
+            a_type = std::get<1>(std::get<0>(p));
+            b_type = std::get<1>(std::get<1>(p));
+          }
```

Comment:
Random suggestion to reduce the redundancy, but I'd structure this like this:

```
  auto a_type = std::get<1>(std::get<0>(p));
  auto b_type = ...
  if (stuff) {
    a_type = a_type.getWidthlessType();
    b_type also
  }
```

Reply:
I opted for a mix of your suggestion and some destructuring logic to unpack things (which I had no idea C++ had / pretty cool!). Let me know if the destructuring is okay to use here.